### PR TITLE
ENH: avoid importing optional dependencies just to know if they are installed

### PR DESCRIPTION
**Pull request by neutrinoceros**
### Description
Inspired by https://github.com/astropy/astropy/issues/13821#issuecomment-1861228534
I noticed that we could define `HAS_<optional_dep>` constant more lazily without breaking tests, though maybe this isn't desired in the overall context of #13821 ?

I bet @mhvk has opinions here.

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

File: astropy/utils/compat/optional_deps.py [hunk 1]
```diff
@@ -44,10 +44,6 @@
 
 def __getattr__(name):
     if name in __all__:
-        try:
-            importlib.import_module(_deps[name[4:]])
-        except (ImportError, ModuleNotFoundError):
-            return False
-        return True
+        return find_spec(name.removeprefix("HAS_").lower()) is not None
```
Comment (mhvk) [inline astropy/utils/compat/optional_deps.py:47]: This should be `find_spec(_deps[name[4:]])`

**Review by mhvk — COMMENTED** [review]
That seems like a clear improvement! (but note small mistake) From a quick test, it looks like `find_spec` is indeed substantially faster.

p.s. I also wondered about doing
```
present = globals()[name] = findspec(_deps[name[4:]]) is not None
return present
```
so that the result is remembered for the next call. But am not sure that saves much time after this!

File: astropy/utils/compat/optional_deps.py (hunk 1 above)
Comment (neutrinoceros) [reply astropy/utils/compat/optional_deps.py:47]: at the risk of bike shedding, isn't `.removeprefix("HAS_")` clearer than `[4:]` (albeit ever so slightly slower) ?

Comment (neutrinoceros) [reply astropy/utils/compat/optional_deps.py]: (though getting the name for `_deps` is clearly required, so I applied that fix)

Comment (mhvk) [reply astropy/utils/compat/optional_deps.py]: Yes, I agree, more obvious what it does.

**Review by mhvk — APPROVED** [review]
Nice improvement, now all good!

**Comment by mhvk** [comment]
Well, I approved too quickly: quite a few failures!

**Comment by neutrinoceros** [comment]
Indeed ! I just learned that `find_spec` *may* raise `ModuleNotFoundError` when called with a *subpackage* name (e.g. `matplotlib.pyplot`). So, the easiest fix from here is
```patch
diff --git a/astropy/utils/compat/optional_deps.py b/astropy/utils/compat/optional_deps.py
index b079a82e99..08140b4439 100644
--- a/astropy/utils/compat/optional_deps.py
+++ b/astropy/utils/compat/optional_deps.py
@@ -37,7 +37,7 @@ _optional_deps = [
 _deps = {k.upper(): k for k in _optional_deps}
 
 # Any subpackages that have different import behavior:
-_deps["PLT"] = "matplotlib.pyplot"
+_deps["PLT"] = "matplotlib"
 
 __all__ = [f"HAS_{pkg}" for pkg in _deps]
 ```
 which seems reasonable to me, since I highly doubt that `matplotlib` will ever *not* ship `matplotlib.pyplot`

**Comment by neutrinoceros** [comment]
pushing the fix now to prove my point. I may be away until Wednesday, so feel free to push/merge/close as needed !

**Comment by mhvk** [comment]
pre-commit.ci autofix

**Comment by mhvk** [comment]
OK, now all tests pass and this looks like a very obvious improvement, so merging!

**Pull request merged**